src/index.ts

```diff
@@ -65,17 +86,72 @@ export function createSelectorCreator<
 >(
   memoize: MemoizeFunction,
   ...memoizeOptionsFromArgs: DropFirst<Parameters<MemoizeFunction>>
+): CreateSelectorFunction<F, MemoizeFunction, typeof defaultMemoize>
+
+export function createSelectorCreator<
+  /** Selectors will eventually accept some function to be memoized */
+  F extends (...args: unknown[]) => unknown,
+  MemoizeFunction extends (func: F, ...options: any[]) => F,
+  ArgsMemoizeFunction extends (
+    func: F,
+    ...options: any[]
+  ) => F = typeof defaultMemoize,
+  /** The additional options arguments to the memoizer */
+  MemoizeOptions extends unknown[] = DropFirst<Parameters<MemoizeFunction>>,
+  ArgsMemoizeOptions extends unknown[] = DropFirst<
+    Parameters<ArgsMemoizeFunction>
+  >
```

Comment:
From the pure TS perspective - this doesn't look quite correct. Those are either **not** generic (if you just want to unpack them from `MemoizeFunction` etc) or they are generic and `MemoizeFunction` etc should **depend** on them (as in: `MemoizeFunction extends (func: F, ...options: MemoizeOptions) => F`)

Reply:
Hmm. My intent is to infer them here (and that's what it's already doing in 4.1).  In other words:

- Look at whatever memoize function was passed in
- Determine the type of its arguments
- Use those

I'm using generics with defaults here as a "precalculate this type once, ahead of time" step, because those get used 2-3 times later on in the file.

Reply:
> I'm using generics with defaults here as a "precalculate this type once, ahead of time" step, because those get used 2-3 times later on in the file.

That technique is fine-ish in isolated internal cases - and it can only give u some improved readability (since u can "alias" something), "precalculation" isn't really a thing that benefits u anyhow because TS implementation is heavily cached. If a type reference receives the same type arguments then the outcome will be the same type (in other words, internally `Foo<{ arg: string }>` is only ever evaluated once, even if you use it in thousand files).

From the pure TS perspective though - those aren't generic in the way you express it. You make this a part of the public API (not quite intentionally), so you allow this:
```ts
createSelectorCreator<AnyFunction, typeof someMemoize, CompletelyUnrelated>(/*...*/)
```